types/src/error.rs

```diff
@@ -16,12 +16,37 @@ impl<T: fmt::Display> fmt::Display for Mismatch<T> {
 	}
 }
 
+/// Invalid params.
+#[derive(Debug)]
+pub struct InvalidParams;
+
+/// Error that may occur when a server fails to execute provided closure/callback.
+#[derive(Debug, thiserror::Error)]
+pub enum ServerCallError {
+	#[error("Invalid params in the RPC call")]
+	/// InvalidParams,
+	InvalidParams(InvalidParams),
+	#[error("Provided context failed: {0}")]
+	/// Provided context/metadata failed.
+	ContextFailed(#[source] Box<dyn std::error::Error + Send + Sync>),
+}
+
+impl From<InvalidParams> for ServerCallError {
+	fn from(params: InvalidParams) -> Self {
+		Self::InvalidParams(params)
+	}
+}
+
 /// Error type.
 #[derive(Debug, thiserror::Error)]
 pub enum Error {
+	/// Error that may occur when a server fails to execute provided closure/callback.
+	/// Note, will most likely imply that the JSON-RPC request was valid.
```

Comment:
Isn't it guaranteed that the request was valid if execution gets to a point where a `ServerCallError` can happen?

Reply:
yes, currently but that could change so I was vague here